quasar-utils/src/main/groovy/net/corda/plugins/quasar/QuasarPlugin.groovy

```diff
@@ -68,49 +69,92 @@ class QuasarPlugin implements Plugin<Project> {
             doFirst {
-                jvmArgs "-javaagent:${project.configurations[QUASAR].singleFile}${extension.options.get()}",
+                jvmArgs "-javaagent:${project.configurations[QUASAR_AGENT].singleFile}${extension.options.get()}",
                         "-Dco.paralleluniverse.fibers.verifyInstrumentation"
             }
         }
         project.tasks.withType(JavaExec).configureEach {
             doFirst {
-                jvmArgs "-javaagent:${project.configurations[QUASAR].singleFile}${extension.options.get()}",
+                jvmArgs "-javaagent:${project.configurations[QUASAR_AGENT].singleFile}${extension.options.get()}",
                         "-Dco.paralleluniverse.fibers.verifyInstrumentation"
             }
         }
     }
 
-    @SuppressWarnings("GrDeprecatedAPIUsage")
-    private static Configuration createRuntimeConfiguration(String name, ConfigurationContainer configurations) {
+    private static Configuration createRuntimeOnlyConfiguration(String name, ConfigurationContainer configurations) {
         Configuration configuration = configurations.findByName(name)
         if (configuration == null) {
-            Configuration parent = configurations.getByName(RUNTIME_CONFIGURATION_NAME)
             configuration = configurations.create(name)
             configuration.transitive = false
-            parent.extendsFrom(configuration)
+            configurations.getByName(RUNTIME_ONLY_CONFIGURATION_NAME).extendsFrom(configuration)
         }
         return configuration
     }
+
+    private static Configuration createCompileOnlyConfiguration(String name, ConfigurationContainer configurations) {
+        Configuration configuration = configurations.findByName(name)
+        if (configuration == null) {
+            configuration = configurations.create(name)
+            configurations.getByName(COMPILE_ONLY_CONFIGURATION_NAME).extendsFrom(configuration)
+            configurations.matching { it.name.endsWith("CompileOnly") }.configureEach { cfg ->
+                cfg.extendsFrom(configuration)
+            }
+        }
+        return configuration
+    }
+
+    private static class QuasarAction implements Action<DependencySet> {
+        private final DependencyHandler handler
+        private final QuasarExtension extension
+        private final boolean isTransitive
+
+        QuasarAction(DependencyHandler handler, QuasarExtension extension, boolean isTransitive) {
+            this.handler = handler
+            this.extension = extension
+            this.isTransitive = isTransitive
+        }
+
+        @Override
+        void execute(DependencySet dependencies) {
+            if (isEmpty(extension.suspendableAnnotation)) {
+                def quasarDependency = handler.create(extension.dependency.get()) { dep ->
+                    dep.transitive = isTransitive
+                }
+                dependencies.add(quasarDependency)
+            }
```

Comment:
shouldn't we add the quasar dependency when `extension.suspendableAnnotation` is **NOT** empty?

Reply:
The idea is that the CorDapp only needs to depend on Quasar for the sake of its `@Suspendable` annotation. So we don't need `quasar-core-osgi` on our compile classpath if we're using something else.

Reply:
ok but that means that by default users will stick to quasar's `@Suspendable` unless they configure manually this plugin to use the `@Suspendable` from `corda-base`, right? My concern is simply they would never switch because:
- default is quasar's `@Suspendable`
- simply switching to `@Suspendable` from `corda-base` would cause their tests to fail unless they remember to also configure the gradle plugin

Reply:
CorDapp developers start from a template Gradle project, which will need updating for Corda 5 anyway. We have never asked them to write their own Gradle projects from scratch. Few people in R3 can even do that.... :wink:.

Reply:
ok